### PR TITLE
[sdk/python] Transitive component dependencies.

### PR DESCRIPTION
**Pull request by pgavlin**
Implement Node/.NET-style dependency semantics for component resources.
Depending on a component implicitly depends on all of the component's
children. The exact set of children depends on exactly when the
component resource is observed.

Part of #7542.

### CODE REVIEW

File: sdk/python/lib/pulumi/runtime/resource.py [hunk 1]
```diff
@@ -661,15 +657,37 @@ def convert_providers(
     return result
 
 
-async def _resolve_depends_on_urns(options: 'ResourceOptions') -> List[str]:
+async def _add_dependency(deps: Set[str], res: 'Resource'):
+    from .. import ComponentResource # pylint: disable=import-outside-toplevel
+
+    if isinstance(res, ComponentResource):
```
Comment (justinvp) [inline sdk/python/lib/pulumi/runtime/resource.py:663]: Should we include a comment along the lines of the following somewhere in the Python SDK?

https://github.com/pulumi/pulumi/blob/e567b4762fe0dbb88ec35db2f3677a95cfdd715b/sdk/nodejs/runtime/resource.ts#L600-L627

**Review by justinvp — APPROVED** [review]
LGTM

File: sdk/python/lib/pulumi/runtime/resource.py [hunk 2]
```diff
@@ -661,15 +657,37 @@ def convert_providers(
...
+        for child in res._childResources:
+            await _add_dependency(deps, child)
+        if not res._remote:
+            return
+
+    urn = await res.urn.future()
+    if urn:
+        print(f"adding dep {urn}")
```
[the remark this answers is not in the document]
Comment (pgavlin) [reply sdk/python/lib/pulumi/runtime/resource.py:671]: Ah, nope

File: sdk/python/lib/pulumi/runtime/resource.py [hunk 3]
```diff
@@ -661,15 +657,37 @@ def convert_providers(
...
 
 
-async def _resolve_depends_on_urns(options: 'ResourceOptions') -> List[str]:
+async def _add_dependency(deps: Set[str], res: 'Resource'):
+    from .. import ComponentResource # pylint: disable=import-outside-toplevel
+
+    if isinstance(res, ComponentResource):
+        for child in res._childResources:
+            await _add_dependency(deps, child)
+        if not res._remote:
```
[the remark this answers is not in the document]
Comment (pgavlin) [reply sdk/python/lib/pulumi/runtime/resource.py:666]: That's correct. We do not add in-process components to the list of dependencies. I'm not sure that there would be any harm in doing so, but this is consistent with Node.

File: sdk/python/lib/pulumi/runtime/resource.py (hunk 1 above)
Comment (pgavlin) [reply sdk/python/lib/pulumi/runtime/resource.py:663]: Yes, that's a good idea.

File: CHANGELOG_PENDING.md [hunk 4]
```diff
@@ -1,49 +1,5 @@
 ### Improvements
 
-- [cli] - Fix the preview experience for unconfigured providers. Rather than returning the
```
Comment (pgavlin) [inline CHANGELOG_PENDING.md:3]: Not sure what's up with the CL. I'll get this fixed before merging.

**Pull request merged**